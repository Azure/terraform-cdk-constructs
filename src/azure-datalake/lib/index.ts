export * from "./filesystem";
export * from "./path";
export * from "./datalake";
