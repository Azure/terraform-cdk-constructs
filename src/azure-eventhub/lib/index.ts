export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
