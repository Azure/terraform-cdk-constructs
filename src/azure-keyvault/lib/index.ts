export * from './certificate'
export * from './key'
export * from './policy'
export * from './secret'
export * from './vault'

