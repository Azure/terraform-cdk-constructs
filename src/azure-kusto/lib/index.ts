export * from "./cluster";
export * from "./compute-specification";
export * from "./database";
