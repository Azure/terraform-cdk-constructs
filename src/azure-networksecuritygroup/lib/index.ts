export * from './preconfigured-rules';
export * from './securitygroup';
