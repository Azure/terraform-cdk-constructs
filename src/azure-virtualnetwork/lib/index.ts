export * from './network';
export * from './peering';
