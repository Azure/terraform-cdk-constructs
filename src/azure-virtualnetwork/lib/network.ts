import { Subnet } from "@cdktf/provider-azurerm/lib/subnet";
import { VirtualNetwork } from "@cdktf/provider-azurerm/lib/virtual-network";
import { Construct } from "constructs";
import { PeerProps, Peer, PeerSettings } from "./peering";
import { AzureResource } from "../../core-azure/lib";

/**
 * Configuration properties for defining a subnet within an Azure Virtual Network.
 */
export interface SubnetConfig {
  /**
   * The name of the subnet. This name must be unique within the context of the virtual network.
   */
  readonly name: string;

  /**
   * A list of address prefixes for the subnet. These are expressed in CIDR notation.
   * For example, '192.168.1.0/24' to define a subnet with a range of IP addresses.
   */
  readonly addressPrefixes: string[];
}

/**
 * Properties for defining an Azure Virtual Network.
 */
export interface NetworkProps {
  /**
   * The name of the resource group under which the virtual network will be created.
   */
  readonly resourceGroupName: string;

  /**
   * Optional: The name of the virtual network. Must be unique within the resource group.
   * If not provided, a default name will be assigned.
   */
  readonly name?: string;

  /**
   * Optional: The Azure region in which to create the virtual network, e.g., 'East US', 'West Europe'.
   * If not specified, the region of the resource group will be used.
   */
  readonly location?: string;

  /**
   * Optional: A list of address spaces for the virtual network, specified in CIDR notation.
   * For example, ['10.0.0.0/16'] defines a single address space. Multiple address spaces can be provided.
   */
  readonly addressSpace?: string[];

  /**
   * Optional: An array of subnet configurations to be created within the virtual network.
   * Each subnet is defined by its name and address prefix(es).
   */
  readonly subnets?: SubnetConfig[];
}

export class Network extends AzureResource {
  readonly props: NetworkProps;
  public readonly name: string;
  public resourceGroupName: string;
  public id: string;
  public readonly virtualNetwork: VirtualNetwork;
  public readonly subnets: { [name: string]: Subnet } = {}; // Map of subnet name to Subnet object

  constructor(scope: Construct, id: string, props: NetworkProps) {
    super(scope, id);

    this.props = props;

    const defaults = {
      name: props.name || `vnet-${this.node.path.split("/")[0]}`,
      location: props.location || "eastus",
      addressSpace: props.addressSpace || ["10.0.0.0/16"],
      subnets: props.subnets || [
        {
          name: "default",
          addressPrefixes: ["10.1.0.0/24"],
        },
      ],
    };

    // Create a virtual network
    const vnet = new VirtualNetwork(this, "vnet", {
      ...defaults,
      resourceGroupName: props.resourceGroupName,
    });

    this.name = vnet.name;
    this.id = vnet.id;
    this.resourceGroupName = vnet.resourceGroupName;
    this.virtualNetwork = vnet;

    // Create subnets within the virtual network
    for (const subnetConfig of defaults.subnets) {
      const subnet = new Subnet(this, subnetConfig.name, {
        name: subnetConfig.name,
        resourceGroupName: props.resourceGroupName,
        virtualNetworkName: vnet.name,
        addressPrefixes: subnetConfig.addressPrefixes,
      });

      this.subnets[subnetConfig.name] = subnet; // Populate the subnetsMap
    }
  }

  // Create Vnet Peering Method
  public addVnetPeering(
    remoteVirtualNetwork: Network,
    localPeerSettings?: PeerSettings,
    remotePeerSettings?: PeerSettings,
  ) {
    const vnetPeerProps: PeerProps = {
      virtualNetwork: this,
      remoteVirtualNetwork: remoteVirtualNetwork,
      localToRemoteSettings: localPeerSettings,
      remoteToLocalSettings: remotePeerSettings,
    };
    new Peer(this, remoteVirtualNetwork.id, vnetPeerProps);
  }
}
