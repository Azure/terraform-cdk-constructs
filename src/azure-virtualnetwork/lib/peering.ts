import { VirtualNetworkPeering } from "@cdktf/provider-azurerm/lib/virtual-network-peering";
import { Construct } from "constructs";
import { Network } from "./network";

/**
 * Interface defining the settings for peer connections.
 */
export interface PeerSettings {
  /**
   * Indicates whether virtual network access is allowed.
   * @default true
   */
  readonly allowVirtualNetworkAccess?: boolean;

  /**
   * Indicates whether forwarded traffic is allowed.
   * @default false
   */
  readonly allowForwardedTraffic?: boolean;

  /**
   * Indicates whether gateway transit is allowed.
   * @default false
   */
  readonly allowGatewayTransit?: boolean;

  /**
   * Indicates whether to use remote gateways.
   * @default false
   */
  readonly useRemoteGateways?: boolean;
}

/**
 * Interface defining the properties for virtual network peerings.
 */
export interface PeerProps {
  /**
   * ID of the local virtual network.
   */
  readonly virtualNetwork: Network;

  /**
   * ID of the remote virtual network.
   */
  readonly remoteVirtualNetwork: Network;

  /**
   * Settings applied from the local virtual network to the remote virtual network.
   */
  readonly localToRemoteSettings: PeerSettings | undefined;

  /**
   * Settings applied from the remote virtual network to the local virtual network.
   */
  readonly remoteToLocalSettings: PeerSettings | undefined;
}

export class Peer extends Construct {
  constructor(scope: Construct, name: string, props: PeerProps) {
    super(scope, name);

    const localtoRemotePeerName =
      props.virtualNetwork.name + "to" + props.remoteVirtualNetwork.name;

    new VirtualNetworkPeering(this, "VNetPeerLocaltoRemote", {
      name: localtoRemotePeerName,
      resourceGroupName: props.virtualNetwork.resourceGroup.name,
      virtualNetworkName: props.virtualNetwork.name,
      remoteVirtualNetworkId: props.remoteVirtualNetwork.id,
      ...props.localToRemoteSettings,
    });

    const remoteToLocalPeerName =
      props.remoteVirtualNetwork.name + "to" + props.virtualNetwork.name;

    new VirtualNetworkPeering(this, "VNetPeerRemotetoLocal", {
      name: remoteToLocalPeerName,
      resourceGroupName: props.remoteVirtualNetwork.resourceGroup.name,
      virtualNetworkName: props.remoteVirtualNetwork.name,
      remoteVirtualNetworkId: props.virtualNetwork.id,
      ...props.remoteToLocalSettings,
    });
  }
}
