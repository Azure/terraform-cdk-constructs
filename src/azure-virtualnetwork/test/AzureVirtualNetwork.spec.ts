import { AzurermProvider } from "@cdktf/provider-azurerm/lib/provider";
import { Testing, TerraformStack } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import { exampleAzureVirtualNetwork } from "./ExampleAzureVirtualNetwork";
import * as vnet from "..";

describe("Azure Virtual Network With Defaults", () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    const app = Testing.app();
    stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureFeature", { features: {} });

    new vnet.Network(stack, "testAzureVirtualNetworkDefaults", {});

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Azure Virtual Network with defaults and checks snapshot", () => {
    expect(Testing.synth(stack)).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});

describe("Azure Virtual Network Example", () => {
  it("renders the Example Azure Virtual Network and checks snapshot", () => {
    expect(
      Testing.synth(
        new exampleAzureVirtualNetwork(
          Testing.app(),
          "testAzureVirtualNetwork",
        ),
      ),
    ).toMatchSnapshot();
  });

  it("check if the produced terraform configuration is valid", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(
      Testing.fullSynth(
        new exampleAzureVirtualNetwork(
          Testing.app(),
          "testAzureVirtualNetwork",
        ),
      ),
    ).toBeValidTerraform();
  });

  it("check if this can be planned", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(
      Testing.fullSynth(
        new exampleAzureVirtualNetwork(
          Testing.app(),
          "testAzureVirtualNetwork",
        ),
      ),
    ).toPlanSuccessfully();
  });
});
