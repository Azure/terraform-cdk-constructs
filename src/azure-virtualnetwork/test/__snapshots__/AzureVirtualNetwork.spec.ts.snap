// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Azure Virtual Network Example renders the Example Azure Virtual Network and checks snapshot 1`] = `
"{
  "output": {
    "resource_group_name": {
      "value": "\${azurerm_resource_group.rg.name}"
    },
    "subnet_name": {
      "value": "\${azurerm_subnet.testAzureVirtualNetworkDefaults_subnet1_2DB834E8.name}"
    },
    "virtual_network_name": {
      "value": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.name}"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_resource_group": {
      "rg": {
        "location": "eastus",
        "name": "rg-\${var.name}"
      }
    },
    "azurerm_subnet": {
      "testAzureRemoteVirtualNetworkDefaults_subnet1_9BF78038": {
        "address_prefixes": [
          "10.1.1.0/24"
        ],
        "name": "subnet1",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3.name}"
      },
      "testAzureRemoteVirtualNetworkDefaults_subnet2_5C103062": {
        "address_prefixes": [
          "10.1.2.0/24"
        ],
        "name": "subnet2",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3.name}"
      },
      "testAzureVirtualNetworkDefaults_subnet1_2DB834E8": {
        "address_prefixes": [
          "10.0.1.0/24"
        ],
        "name": "subnet1",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.name}"
      },
      "testAzureVirtualNetworkDefaults_subnet2_5D8D2CC1": {
        "address_prefixes": [
          "10.0.2.0/24"
        ],
        "name": "subnet2",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.name}"
      }
    },
    "azurerm_virtual_network": {
      "testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3": {
        "address_space": [
          "10.1.0.0/16"
        ],
        "location": "westus",
        "name": "vnet-\${var.name}2",
        "resource_group_name": "\${azurerm_resource_group.rg.name}"
      },
      "testAzureVirtualNetworkDefaults_vnet_D8C7A0F3": {
        "address_space": [
          "10.0.0.0/16"
        ],
        "location": "eastus",
        "name": "vnet-\${var.name}",
        "resource_group_name": "\${azurerm_resource_group.rg.name}"
      }
    },
    "azurerm_virtual_network_peering": {
      "testAzureVirtualNetworkDefaults_TfTokenTOKEN36_VNetPeerLocaltoRemote_F643C62C": {
        "name": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.name}to\${azurerm_virtual_network.testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3.name}",
        "remote_virtual_network_id": "\${azurerm_virtual_network.testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3.id}",
        "resource_group_name": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.resource_group_name}",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.name}"
      },
      "testAzureVirtualNetworkDefaults_TfTokenTOKEN36_VNetPeerRemotetoLocal_F5E01733": {
        "name": "\${azurerm_virtual_network.testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3.name}to\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.name}",
        "remote_virtual_network_id": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.id}",
        "resource_group_name": "\${azurerm_virtual_network.testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3.resource_group_name}",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureRemoteVirtualNetworkDefaults_vnet_638C5DB3.name}"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.75.0"
      }
    }
  },
  "variable": {
    "name": {
      "default": "test",
      "description": "System name used to randomize the resources",
      "type": "string"
    }
  }
}"
`;

exports[`Azure Virtual Network With Defaults renders an Azure Virtual Network with defaults and checks snapshot 1`] = `
"{
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_subnet": {
      "testAzureVirtualNetworkDefaults_default_85452F83": {
        "address_prefixes": [
          "10.1.0.0/24"
        ],
        "name": "default",
        "resource_group_name": "rg-test",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureVirtualNetworkDefaults_vnet_D8C7A0F3.name}"
      }
    },
    "azurerm_virtual_network": {
      "testAzureVirtualNetworkDefaults_vnet_D8C7A0F3": {
        "address_space": [
          "10.0.0.0/16"
        ],
        "location": "eastus",
        "name": "vnet-test",
        "resource_group_name": "rg-test"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.75.0"
      }
    }
  }
}"
`;
